src/client/components/date-range-picker/date-range-picker.tsx

```diff
@@ -0,0 +1,225 @@
+require('./date-range-picker.css');
+
+import * as React from "react";
+import { Timezone, WallTime, Duration, minute, hour, day, week, month, year } from "chronoshift";
+import { TimeRange } from "plywood";
+import { formatFullMonthAndYear, prependDays, appendDays, daysEqualWallTime,
+  getEndWallTimeInclusive, monthToWeeks, wallTimeHelper } from "../../utils/date/date";
+import { classNames } from "../../utils/dom/dom";
+import { getLocale } from "../../config/constants";
+import { SvgIcon } from "../svg-icon/svg-icon";
+import { DateInput } from "../date-input/date-input";
+
+
+export interface DateRangePickerProps extends React.Props<any> {
+  startTime?: Date;
+  endTime?: Date;
+  maxTime?: Date;
+  timezone: Timezone;
+  onStartChange: (t: Date) => void;
+  onEndChange: (t: Date) => void;
+}
+
+export interface DateRangePickerState {
+  activeMonthStartDate?: Date;
+  hoverTimeRange?: TimeRange;
+  selectionSet?: boolean;
+}
+
+export class DateRangePicker extends React.Component<DateRangePickerProps, DateRangePickerState> {
+  constructor() {
+    super();
+    this.state = {
+      activeMonthStartDate: null,
+      hoverTimeRange: null,
+      selectionSet: false
+    };
+  }
+
+  componentWillMount() {
+    var { startTime } = this.props;
+    if (!startTime) return;
+    if (isNaN(startTime.valueOf())) {
+      this.setState({
+        activeMonthStartDate: null
+      });
+      return;
+    }
+
+    this.setState({
+      activeMonthStartDate: startTime
+    });
+  }
+
+  navigateToMonth(offset: number): void {
+    const { timezone } = this.props;
+    const { activeMonthStartDate } = this.state;
+    var newDate = month.shift(activeMonthStartDate, timezone, offset);
+    this.setState({
+      activeMonthStartDate: newDate
+    });
+  }
+
+  goToPreviousMonth(): void {
+    return this.navigateToMonth(-1);
+  }
+
+  goToNextMonth(): void {
+    return this.navigateToMonth(1);
+  }
+
+  calculateHoverTimeRange(mouseEnteredDay: Date) {
+    const { startTime, endTime } = this.props;
+    var hoverTimeRange: TimeRange = null;
+    if (startTime && !endTime) {
+      var start = startTime;
+      var end = mouseEnteredDay;
+      // if mousing over backwards, set end to old start time
+      if (mouseEnteredDay < startTime) {
+        start = mouseEnteredDay;
+        end = startTime;
+      }
+      hoverTimeRange = new TimeRange({ start, end });
+    }
+
+    this.setState({ hoverTimeRange });
+  }
+
+  selectDay(selection: Date): void {
+    const { onStartChange, onEndChange, startTime, endTime, timezone } = this.props;
+    const tzString = timezone.toString();
+    if (!startTime) return;
+    var selectionFloored = day.floor(selection, timezone);
+
+    if (endTime) {
+      // clicking on end date of range should do nothing
+      if (daysEqualWallTime(selection, getEndWallTimeInclusive(endTime, timezone), tzString)
```

Comment:
It could be more explicit to put those conditions into small functions with self-descripting names, thus avoiding too the need for comments.